### PR TITLE
Minor typescript organization

### PR DESCRIPTION
**Pull request by thebearingedge**
* toJSON() of JSONValue returns JSONValue type
* TTypes defaults to JSToPostgresTypMap
* JSToPostgresTypeMap is in postgres namespace

The JSONValue type already includes recursive type references, so we may as well
guarantee that JSON.stringify() will get a JSONValue back.

Having to always include {} when referencing Sql<{}> is annoying.

This exposes the type parameter of the Sql interface for extension by other libraries that would like to extend it.

### CODE REVIEW

**Comment by porsager** [comment]
Would you mind taking a look @Minigugus ?

File: types/index.d.ts
Comment (Minigugus) [inline types/index.d.ts:479]: Types are already quite strict in this library, and enforcing types for `toJSON()` may cause troubles when integrating with others libraries that don't respect the `JSONValue` type definition. Anyway, it's not `postgres` responsibility to validate user-side `toJSON` implementations: users that care about proper types should type the `toJSON` definition instead.

```suggestion
    | { toJSON(): any } // `toJSON` called by `JSON.stringify`; not typing the return type, types definition is strict enough anyway
```

**Review by Minigugus — CHANGES_REQUESTED** [review]

File: types/index.d.ts
Comment (thebearingedge) [reply types/index.d.ts:584]: No problem 👌

File: types/index.d.ts
Comment (thebearingedge) [reply types/index.d.ts:479]: Ok, I think that's fair 👍

**Comment by thebearingedge** [comment]
@Minigugus Thank you for the review!

**Review by Minigugus — CHANGES_REQUESTED** [review]
~~I guess you've tried to rebase or something? :thinking:~~

Ok you just ran the `build` script, sorry :sweat_smile: Maybe it would be better to either commit the `build` script result before merging, or at least only push types changes?

**Comment by thebearingedge** [comment]
@Minigugus Sorry, I believe that the build script ran due to the `prepare` script, but I am not accustomed to committing build artifacts so I didn't explicitly add it.

I'm pretty sure I botched synchronizing my fork with upstream, hence the rebase. I am going to attempt to clean things up this morning.

**Comment by Minigugus** [comment]
@thebearingedge Do you think you could change `PostgresTypeList` to `Record<string, postgres.PostgresType>` too? It's quite similar to the fix for `JSToPostgresTypeMap` so it is relevant to fix `PostgresTypeList` here too? #448 would be solved by this issue then

**Review by RealShadowNova — APPROVED** [review]

**Comment by thebearingedge** [comment]
@Minigugus yes, I should be able to take care of that 👍

File: types/index.d.ts
Comment (thebearingedge) [inline types/index.d.ts:129]: I have updated the `PostgresTypeList` to extend a `Record`.

File: types/index.d.ts
Comment (thebearingedge) [inline types/index.d.ts:18]: I am happy to change this, but I could not figure out why there was a condition for `PostgresTypeList extends T` and when I tried my fork it seemed to have the same effect when omitted.

**Review by thebearingedge — COMMENTED** [review]
@Minigugus I made a modification to the signatures of `postgres<T>`, but am happy to undo it. There may be some reason for the condition that I am unaware of.

File: types/index.d.ts
Comment (Minigugus) [reply types/index.d.ts:129]: The issue with `PostgresTypeList` is that it's confusing for many users, so the best would be to just replace this type with `Record<string, postgres.PostgresType>`, just like you did with `JSToPostgresTypeMap` :+1:

File: types/index.d.ts
Comment (Minigugus) [reply types/index.d.ts:18]: This is because sometimes TS infer `PostgresTypeList` (and not `{}`) under some conditions, then the `keyof T` is replace with `string` in the following object literal, and the `TTypes` generic become just a `Record<string, CustomTypeA | CustomTypeB | ...>`, which is quite useless. But maybe recent versions of typescript have fixed that, I checked again recently

**Comment by thebearingedge** [comment]
@Minigugus Ok! I think we are all set 🔥 ⌨️ 🔥

**Comment by porsager** [comment]
Is this ready to merge once conflicts have been resolved @Minigugus ?

**Review by Minigugus — APPROVED** [review]
Sorry for the late reply

Looks good to me :+1:

**Comment by porsager** [comment]
Would any of you mind resolving the current conflicts ? :)

**Comment by thebearingedge** [comment]
Yes I will tackle it sometime this week 👍

**Comment by thebearingedge** [comment]
@porsager thank you, the conflict has been resolved :salute:

File: types/index.d.ts
Comment (Minigugus) [inline types/index.d.ts:8]:
```suggestion
declare function postgres<T extends Record<string, postgres.PostgresType> = {}>(options?: postgres.Options<T> | undefined): postgres.Sql<Record<string, postgres.PostgresType> extends T ? {} : { [type in keyof T]: T[type] extends {
```

The `| undefined` part is on purpose: #407

File: types/index.d.ts
Comment (Minigugus) [inline types/index.d.ts:18]:
```suggestion
declare function postgres<T extends Record<string, postgres.PostgresType> = {}>(url: string, options?: postgres.Options<T> | undefined): postgres.Sql<Record<string, postgres.PostgresType> extends T ? {} : { [type in keyof T]: T[type] extends {
```

Same here

**Review by Minigugus — APPROVED** [review]
Just add theses missing `undefined` and it's ready to merge :slightly_smiling_face:

**Comment by akatechis** [comment]
Would this PR fix the error `type "string" does not exist`? I'm using postgresjs with the following code example:

```typescript
import postgres from 'https://deno.land/x/postgresjs@v3.1.0/mod.js';

const sql = postgres<PGTypes>({
  host: 'localhost',
  port: 5432,
  database: 'postgres',
  user: 'postgres',
  password: 'postgres',
});

interface PGTypes {
  string: 'text';
  datetime: 'timestamp';
  integer: 'int';
  [name: string]: string;
}
```

But when I run it, I get the following error, and something tells me I'm not passing the correct `JSToPostgresTypeMap` type parameter to the `postgres` factory function...

```
error: Uncaught (in promise) PostgresError: type "string" does not exist
    const error = Errors.postgres(parseError(x))
                         ^
    at ErrorResponse (https://deno.land/x/postgresjs@v3.1.0/src/connection.js:761:26)
    at handle (https://deno.land/x/postgresjs@v3.1.0/src/connection.js:468:6)
    at data (https://deno.land/x/postgresjs@v3.1.0/src/connection.js:312:9)
    at https://deno.land/x/postgresjs@v3.1.0/polyfills.js:140:32
    at Array.forEach (<anonymous>)
    at call (https://deno.land/x/postgresjs@v3.1.0/polyfills.js:140:18)
    at success (https://deno.land/x/postgresjs@v3.1.0/polyfills.js:100:11)
```

If there's any documentation on what to pass here, I would gladly accept a link so I can read up on how to pass a correct generic interface

**Comment by thebearingedge** [comment]
Hi @akatechis, The `types` option will need to be an object, like this example under Custom Types.

https://github.com/porsager/postgres#custom-types

**Comment by akatechis** [comment]
Thanks, @thebearingedge, i'll look into that. Does that require me to give type definitions for basic types like string as well, or is that just for custom types? The table I'm trying to query is just two string columns, created by the following:

```typescript
sql`create table if not exists ${sql(tableName)} (id string not null primary key, data string)`
```

I would assume I don't need to define how to handle strings. Is that correct?

**Comment by thebearingedge** [comment]
@akatechis no worries 😄 

regarding your custom `string` type. for that you will need to define it as a domain, as i do not believe that postgres has it. there is a `text` type though.

```sql
create domain string as text;
```

https://www.postgresql.org/docs/current/sql-createdomain.html

the custom type settings for the `postgres` package are there mainly to let you pre-process and post-process values transferring to/from your JavaScript code. so you might have something like this (just a random example)

```ts
const types = {
  string: {
    to: 25, // the PG oid for "text"
    from: [25],
    serialize: s => s.toLowerCase(), // lowercase before storing in PG
    parse: s => s.toUpperCase()  // uppercase before using in JS
  }
}
```

**Pull request merged**
